migrateFromSessionToken: incorrect handling of network errors

STR:
- Perform a `migrateFromSessionToken`, have it fail due to a network error
```
         FirefoxAccount  I  Executing: migrateFromSessionToken
       libfxaclient_ffi  D  fxaclient_ffi: fxa_migrate_from_session_token
                         I  viaduct::backend: Using HTTP backend reqwest (untrusted)
                         D  hyper::client::connect::dns: resolving host="accounts.firefox.com"
                         E  viaduct::backend::reqwest: Reqwest error: Error(Io(Custom { kind: TimedOut, error: StringError("timed out") }), "https://accounts.firefox.com/.well-known/fxa-client-configuration")
                         W  fxa_client::ffi: Network error: Error(
                         W  Network error: Network error: https://accounts.firefox.com/.well-known/fxa-client-configuration: timed out)
         FirefoxAccount  E  Error while running: migrateFromSessionToken
                         E  mozilla.appservices.fxaclient.FxaException$Network: Network error: Network error: https://accounts.firefox.com/.well-known/fxa-client-configuration: timed out
                         E      at mozilla.appservices.fxaclient.rust.RustError.intoException(RustError.kt:43)
                         E      at mozilla.appservices.fxaclient.FirefoxAccount.migrateFromSessionToken(FirefoxAccount.kt:664)
                         E      at mozilla.components.service.fxa.FirefoxAccount$migrateFromSessionTokenAsync$1$1.invoke(FirefoxAccount.kt:143)
                         E      at mozilla.components.service.fxa.FirefoxAccount$migrateFromSessionTokenAsync$1$1.invoke(FirefoxAccount.kt:28)
                         E      at mozilla.components.service.fxa.UtilsKt$handleFxaExceptions$5.invoke(Utils.kt:65)
                         E      at mozilla.components.service.fxa.UtilsKt$handleFxaExceptions$5.invoke(Unknown Source:0)
                         E      at mozilla.components.service.fxa.UtilsKt.handleFxaExceptions(Utils.kt:31)
                         E      at mozilla.components.service.fxa.UtilsKt.handleFxaExceptions(Utils.kt:57)
                         E      at mozilla.components.service.fxa.UtilsKt.handleFxaExceptions(Utils.kt:64)
                         E      at mozilla.components.service.fxa.FirefoxAccount$migrateFromSessionTokenAsync$1.invokeSuspend(FirefoxAccount.kt:14
                            2)
                         E      at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                         E      at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:238)
                         E      at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
                         E      at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
                         E      at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:742)
```
- Then run `migrateFromSessionToken` again, and observe that internally `sessionToken` was already set and remained as set, even though the previous call didn't succeed. At this point we're in a bad state.
```
         FirefoxAccount  I  Executing: migrateFromSessionToken
       libfxaclient_ffi  D  fxaclient_ffi: fxa_migrate_from_session_token
                         W  fxa_client::ffi: Unexpected error: Error(
                         W  Illegal state: Session Token is already set.)

         FirefoxAccount  E  Error while running: migrateFromSessionToken
                         E  mozilla.appservices.fxaclient.FxaException$Unspecified: Illegal state: Session Token is already set.
                         E      at mozilla.appservices.fxaclient.rust.RustError.intoException(RustError.kt:48)
                         E      at mozilla.appservices.fxaclient.FirefoxAccount.migrateFromSessionToken(FirefoxAccount.kt:664)
                         E      at mozilla.components.service.fxa.FirefoxAccount$migrateFromSessionTokenAsync$1$1.invoke(FirefoxAccount.kt:143)
                         E      at mozilla.components.service.fxa.FirefoxAccount$migrateFromSessionTokenAsync$1$1.invoke(FirefoxAccount.kt:28)
                         E      at mozilla.components.service.fxa.UtilsKt$handleFxaExceptions$5.invoke(Utils.kt:65)
                         E      at mozilla.components.service.fxa.UtilsKt$handleFxaExceptions$5.invoke(Unknown Source:0)
                         E      at mozilla.components.service.fxa.UtilsKt.handleFxaExceptions(Utils.kt:31)
                         E      at mozilla.components.service.fxa.UtilsKt.handleFxaExceptions(Utils.kt:57)
                         E      at mozilla.components.service.fxa.UtilsKt.handleFxaExceptions(Utils.kt:64)
                         E      at mozilla.components.service.fxa.FirefoxAccount$migrateFromSessionTokenAsync$1.invokeSuspend(FirefoxAccount.kt:142)
                         E      at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                         E      at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:238)
                         E      at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
                         E      at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
                         E      at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:742)
```

--- a/components/fxa-client/src/migrator.rs
+++ b/components/fxa-client/src/migrator.rs
@@ -30,7 +30,7 @@ impl FirefoxAccount {
             .duplicate_session(&self.state.config, &session_token)?;
 
         let duplicated_session_token = duplicate_session.session_token;
-        self.state.session_token = Some(duplicated_session_token.clone());
+
         let oauth_response = self.client.oauth_tokens_from_session_token(
             &self.state.config,
             &duplicated_session_token,
@@ -58,6 +58,7 @@ impl FirefoxAccount {
             k: k_sync,
             kid,
         };
+        self.state.session_token = Some(duplicated_session_token.clone());
         self.state
             .scoped_keys
             .insert(scopes::OLD_SYNC.to_string(), k_sync_scoped_key);

